Strip micro-version and add upper bound when adding package imports

alshamams opened the pull request:
 This commit resets the micro-version to ‘0’ while importing packages as part of quick fix, as it is not relevant. 

Fixes: https://github.com/eclipse-pde/eclipse.pde/issues/401

HannesWell reviewed (changes requested):
Hi @alshamams thank you for working on this.
While this fixes the linked issue in its most basic form I think this is a good opportunity to fix a few other flaws in this quick-fix.

In general it would be good to also specify the upper bound of the import version range and to move the logic to a more general place, i.e. to `ImportPackageObject.getVersion()`.
You can get such String using the following to assemble a suitable VersionRange from a `org.osgi.framework.Version`:
```
new VersionRange("[" + version.getMajor() + "." + version.getMinor() + "," + (version.getMajor() + 1) + "]").toString()
```
With that the version of the `ImportPackageObject` does not have to be set explicitly anymore.

Besides that, the quick-fix currently only adds a version if the Import-Package header already exists. This could be fixed by replacing.

```
if (header == null) {
	bundle.setHeader(Constants.IMPORT_PACKAGE, pkgId);
} else if (header instanceof ImportPackageHeader ipHeader) {
```
by
```
if (header == null) {
	header = bundle.getModel().getFactory().createHeader(Constants.IMPORT_PACKAGE, pkgId);
}
if (header instanceof ImportPackageHeader ipHeader) {
```

alshamams commented:
Thanks @HannesWell for the quick review.

> In general it would be good to also specify the upper bound of the import version range and to move the logic to a more general place

Thanks, this looks clean and elegant. However, modifying the getVersion API will have implications in regular import package flow. (Not the quick fix one, but while adding import packages from Dependencies tab). Is this what you intend ? Just wanted to make sure you are aware of this change before I push.

> Besides that, the quick-fix currently only adds a version if the Import-Package header already exists. This could be fixed by replacing.

I was aware of it, and was planning to do it in a follow-up PR, given that it is unrelated to the issue at hand. I think that I will make the change here itself.

HannesWell commented:
> > In general it would be good to also specify the upper bound of the import version range and to move the logic to a more general place
> 
> Thanks, this looks clean and elegant. However, modifying the getVersion API will have implications in regular import package flow. (Not the quick fix one, but while adding import packages from Dependencies tab). Is this what you intend ? Just wanted to make sure you are aware of this change before I push.

I have not yet checked every caller path of the code, but yes in general it is good practice to specifiy an upper bound if one specifies a lower to follow semantic versioning. If one really wants to have no upper-bound, it can still be removed later. But from my experience I currently add an upper bound in 90% of the time I use the quick-fix.

> 
> > Besides that, the quick-fix currently only adds a version if the Import-Package header already exists. This could be fixed by replacing.
> 
> I was aware of it, and was planning to do it in a follow-up PR, given that it is unrelated to the issue at hand. I think that I will make the change here itself.

Yes that's fine from my POV, this change is small enough at the moment and it is related enough to mix it.

alshamams commented:
Hi @HannesWell ,
I have addressed the changes, and squashed and rebased the commits while adding you as the co-author. Kindly let me know in case of any more edits.

HannesWell commented on ui/org.eclipse.pde.core/text/org/eclipse/pde/internal/core/text/bundle/ImportPackageObject.java:
```diff
@@ -32,9 +32,17 @@ public class ImportPackageObject extends PackageObject {
-		String version = desc.getVersion().toString();
-		if (!version.equals(Version.emptyVersion.toString())) {
-			return desc.getVersion().toString();
+		// String version = desc.getVersion().toString();
+		//
+		// if (!version.equals(Version.emptyVersion.toString())) {
+		// return desc.getVersion().toString();
+		// }
+		// return null;
```
Please delete commented out code.
```suggestion
```

HannesWell commented on ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/correction/java/JavaResolutionFactory.java:
```diff
@@ -321,14 +321,18 @@ protected void modifyModel(IBaseModel model, IProgressMonitor monitor) throws Co
+					if (header instanceof ImportPackageHeader ipHeader) {
 						int manifestVersion = BundlePluginBase.getBundleManifestVersion(bundle);
-						String versionAttr = (manifestVersion < 2) ? ICoreConstants.PACKAGE_SPECIFICATION_VERSION : Constants.VERSION_ATTRIBUTE;
-						ImportPackageObject impObject = new ImportPackageObject((ManifestHeader) header, (ExportPackageDescription) getChangeObject(), versionAttr);
+						String versionAttr = (manifestVersion < 2) ? ICoreConstants.PACKAGE_SPECIFICATION_VERSION
+								: Constants.VERSION_ATTRIBUTE;
+						ImportPackageObject impObject = new ImportPackageObject((ManifestHeader) header, desc,
+								versionAttr);
```
If you inline `manifestVersion` and replace header by `ipHeader` we would get some nicely formatted code.
```suggestion
						String versionAttr = (BundlePluginBase.getBundleManifestVersion(bundle) < 2)
								? ICoreConstants.PACKAGE_SPECIFICATION_VERSION
								: Constants.VERSION_ATTRIBUTE;
						ImportPackageObject impObject = new ImportPackageObject(ipHeader, desc, versionAttr);
```

HannesWell reviewed (approved):
@alshamams I just applied the suggestions locally and also added handling for the default empty version (`0.0.0`) in `ImportPackageObject` so that in that case not a version range like `[0.0.0,1.0.0)` is added. This case 

The great thing about specifying the upper bound in `ImportPackageObject.getVersion()` is, that it is now also added if one select and adds a package via the Editor directly or through the `add dependencies` link in the `Automated Management of Dependencies` section.
This is IMHO a great improvement for PDE respectively its users.

The current state looks very good to me and is for me ready for submission.
Are you fine with this as well?

HannesWell commented:
I'm merging this now.
Thank you again @alshamams.
